### PR TITLE
ui: Use "channel" or "stream" based on server feature level

### PR DESCRIPTION
**Pull request by chrisbobbe**
Done at the level of TranslationProvider, which (conveniently) can
access the feature level of the active account if any.

Fixes: #5827

### CODE REVIEW

File: src/boot/streamChannelRenamesMap.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ * instead of
+ *
+ *   "Notify stream": "Notify stream",
+ *
+ * and likewise for all the other languages.
+ */
+export const streamChannelRenamesMap: {| [string]: string |} = {
+  stream: 'channel',
+  'Notify stream': 'Notify channel',
+  'Who can access the stream?': 'Who can access the channel?',
```
[the remark this answers is not in the document]
Comment (gnprice) [reply src/boot/streamChannelRenamesMap.js:32]: It also occurs to me that that `{streamAndTopic}` string might be worth including but commented out, as a bit of a reminder of why this map isn't just a `replaceAll` or the like.

**Review by gnprice — COMMENTED** [review]
Thanks @chrisbobbe for taking care of this! Glad to see this is a lot cleaner (thanks to the idea you described on Friday and implemented here) than I was originally imagining.

The changes all look good modulo the TODOs you have, and a few small things below.

File: static/translations/messages_en.json [hunk 2]
```diff
@@ -15,8 +15,10 @@
   "all": "all",
   "everyone": "everyone",
   "stream": "stream",
+  "channel": "channel",
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply static/translations/messages_en.json:18]: Cool; thanks!

File: src/autocomplete/WildcardMentionItem.js [hunk 3]
```diff
@@ -38,14 +41,17 @@ export enum WildcardMentionType {
 // All of these should appear in messages_en.json so we can make the
 // wildcard mentions discoverable in the people autocomplete in the client's
 // own language. See getWildcardMentionsForQuery.
-const englishCanonicalStringOf = (type: WildcardMentionType): string => {
+const englishCanonicalStringOf = (
+  type: WildcardMentionType,
+  channelTerminologyUsed: boolean,
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/autocomplete/WildcardMentionItem.js:46]: Interesting; yeah, I agree. Thanks!

File: src/autocomplete/WildcardMentionItem.js [hunk 4]
```diff
@@ -135,9 +150,12 @@ export default function WildcardMentionItem(props: Props): Node {
 
   const _ = useContext(TranslationContext);
 
+  const zulipFeatureLevel = useSelector(getZulipFeatureLevel);
+  const channelTerminologyUsed = zulipFeatureLevel >= streamChannelRenameFeatureLevel;
+
   const handlePress = useCallback(() => {
-    onPress(type, serverCanonicalStringOf(type));
-  }, [onPress, type]);
+    onPress(type, serverCanonicalStringOf(type, channelTerminologyUsed));
```
[the remark this answers is not in the document]
Comment (chrisbobbe) [reply src/autocomplete/WildcardMentionItem.js:157]: Ah yeah, makes sense. I started a discussion in `#api design`: https://chat.zulip.org/#narrow/stream/378-api-design/topic/stream.2Fchannel.20rename.3A.20.40stream.20wildcard.20mention/near/1752174

**Comment by chrisbobbe** [comment]
Thanks for the review! Revision pushed.

**Comment by chrisbobbe** [comment]
And converted to draft, to help prevent accidentally merging this without doing the TODOs in the commit message.

**Comment by gnprice** [comment]
Thanks! All looks good modulo those TODOs.

I think you can go ahead and fill in the todo-server comments with a reference to 9.0. This is one issue that seems pretty solidly like we're going to make it part of the 9.0 release. But then the other two just mean we're waiting for the server-side changes to be ready before we can merge.

I guess one thing that would be good, since we're waiting, is to get the "channel" strings into Transifex so that people can translate them whenever they get to doing some Zulip translation. So perhaps pull out a commit that just adds to messages_en.json without yet adding the logic that uses the new strings, and then we can merge that and do tools/tx-sync.

**Comment by chrisbobbe** [comment]
Makes sense! Done; revision pushed.

**Comment by gnprice** [comment]
Thanks! All LGTM.

I'll go ahead and merge that first commit:
50c9656a5 i18n: Offer UI strings with "channel" terminology to translators
and run `tools/tx-sync`.

Then the remaining changes are just waiting for the server-side changes to be ready, so that we know what feature level to condition on (and can confirm that accepting `@**channel**` happens before then, so that just one threshold feature level suffices).

**Comment by gnprice** [comment]
The server change to accept `@**channel**` was merged:
* https://github.com/zulip/zulip/pull/29311

So that will indeed happen before the other server-side changes; and this PR is now just waiting for those other server-side changes to land so we can determine the feature-level threshold.

**Comment by gnprice** [comment]
The server-side UI changes have landed — the feature level to use is 255:
https://chat.zulip.org/api/changelog#changes-in-zulip-90

So this is now ripe to update for merge.

(It looks like the main commit's commit message has a couple of elements that would have belonged instead in the commit 50c9656a5cfea9c06c36901e0f39ddd888e9bd10 that we merged earlier; ah well.)

**Comment by chrisbobbe** [comment]
> (It looks like the main commit's commit message has a couple of elements that would have belonged instead in the commit [50c9656](https://github.com/zulip/zulip-mobile/commit/50c9656a5cfea9c06c36901e0f39ddd888e9bd10) that we merged earlier; ah well.)

Ah yeah, the Perl command?

**Comment by chrisbobbe** [comment]
Thanks for the bump! Revision pushed, and I've unmarked as draft.

**Comment by gnprice** [comment]
Thanks! LGTM; merging.

> Ah yeah, the Perl command?

Yeah, exactly — the bit about updating `messages_en.json`.

**Pull request merged**